examples/cpu/ccl_example.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "edm/cell.hpp"
+#include "algorithms/component_connection.hpp"
+#include "csv/csv_io.hpp"
+
+#include "vecmem/memory/host_memory_resource.hpp"
+
+#include <iostream>
+#include <algorithm>
+#include <iomanip>
+#include <chrono>
+
+namespace {
+    double delta_ms(
+        std::chrono::high_resolution_clock::time_point s,
+        std::chrono::high_resolution_clock::time_point e
+    ) {
+        return std::chrono::duration_cast<std::chrono::microseconds>(e - s).count() / 1000.0;
```

Comment:
This is only meant as an "educational corner" here, for this executable I don't care about this too much.

You should **never** perform a division with a constant value in your code. You should always rather perform a multiplication with the inverse of the constant. So in this case you would rather write:

```c++
return std::chrono::duration_cast<std::chrono::microseconds>(e - s).count() * 1E-3;
```

In more complicated cases you can leave it up to the compiler to perform the division at compile time by using `constexpr` expressions.

As I started, this is not really relevant here. But in other parts of this project we should all be very aware of how much slower divisions are than multiplications. :wink:

Reply:
I don't know. That's like calling the passing of objects by references instead of by value (in a place where passing by reference is appropriate) a premature optimisation. :thinking: In any numerical calculation not dividing by a constant should be as much of a basic code writing pattern as any other.

And this is why I wanted to bring it up on this example, as an educational comment.

Reply:
I'm with you that, generally, in performance-sensitive code, it's a good idea to multiply by the reciprocal rather than divide, but I think a blanket statement that division is _always_ bad is actually quite harmful. There's a few reasons why one might want to divide rather than multiply. The most obvious, of course, is numerical accuracy. There is a very good reason why compilers will not (by default) make the transformation between `x / N` to `x * (1.0 / N)`: while it's valid in "proper" mathematics, the two are not equivalent in IEEE 754 arithmetic. Sure, the difference is usually not huge, and the result is biased either way, but the fact of the matter is that the two approaches will give you two different results. Whether the difference is acceptable is application-specific, and usually in scientific computing it's fine. And then there is the problem of different handling of subnormal numbers and other such IEEE 754 oddities.

Writing code using division is often a lot clearer than code written using multiplication with the reciprocal, because the programmer's intent is immediately clear. Reciprocal multiplication usually involves declaring additional variable, and it makes the code harder to read. I'll actually go out on a limb here and say that if you're going to be lax in your IEEE compliance anyway you might as well turn on `-funsafe-math-optimizations`, at least locally: that will leave your code in a more readable state and the compiler will probably be able to do a better job than most humans anyway. In this particular case, we're working with nice, round, powers of 10, so the two approaches are equivalent in readability. But in general, I find code written in the form `a / 27.0` much more readable than `a * 0.037037037`.

And most importantly, making this transformation is not even really guaranteed to speed up your code. If your kernel is memory-bound, for example, a transformation like really just makes your code harder to read, potentially less numerically accurate, for no discernible speed gain.

Just to be clear, once again, that you are absolutely correct that replacing division by multiplication is _often_ a good idea (although I would argue to let the compiler do it for you). I am just not a fan of making blanket statements like this. :slightly_smiling_face: